packages/did-comm/src/utils.ts

```diff
@@ -0,0 +1,306 @@
+}
+
+export function base64ToBytes(s: string): Uint8Array {
+  const inputBase64Url = s.replace(/\+/g, '-').replace(/\//g, '_').replace(/=/g, '')
+  return u8a.fromString(inputBase64Url, 'base64url')
+}
+
+export function bytesToBase64(b: Uint8Array): string {
+  return u8a.toString(b, 'base64pad')
+}
+
+export function encodeBase64url(s: string): string {
+  return bytesToBase64url(u8a.fromString(s))
+}
+
+export function decodeBase64url(s: string): string {
+  return u8a.toString(base64ToBytes(s))
+}
+
+export type DIDCommPlainMessage = IDIDCommMessage & { typ: DIDCommMessageMediaType.PLAIN }
+
+export type DIDCommEncryptedMessage = JWE
+
+export type FlattenedJWS = {
+  payload: string
+  protected?: string
+  header?: Record<string, any>
+  signature: string
+}
+
+export type GenericJWS = {
+  payload: string
+  signatures: [{ protected?: string; header?: Record<string, any>; signature: string }]
+}
+
+export type DIDCommSignedMessage = FlattenedJWS | GenericJWS
+
+export type ExtendedVerificationMethod = VerificationMethod & { publicKeyBase64?: string }
+
+/**
+ * represents an IKey that has been augmented with its corresponding entry from a DID document
+ *
+ * this is only used internally
+ */
+export interface ExtendedIKey extends IKey {
+  meta: KeyMetadata & {
+    verificationMethod: NormalizedVerificationMethod
+  }
+}
+
+/**
+ * represents a VerificationMethod whose public key material has been converted to publicKeyHex
+ */
+export type NormalizedVerificationMethod = Omit<
+  VerificationMethod,
+  'publicKeyBase58' | 'publicKeyBase64' | 'publicKeyJwk'
+>
+
+export function encodeJoseBlob(payload: {}) {
+  return u8a.toString(u8a.fromString(JSON.stringify(payload), 'utf-8'), 'base64url')
+}
+
+export function decodeJoseBlob(blob: string) {
+  return JSON.parse(u8a.toString(u8a.fromString(blob, 'base64url'), 'utf-8'))
+}
+
+export function isDefined<T>(arg: T): arg is Exclude<T, null | undefined> {
+  return typeof arg !== 'undefined'
+}
+
+export function createEcdhWrapper(secretKeyRef: string, context: IAgentContext<IKeyManager>): ECDH {
+  return async (theirPublicKey: Uint8Array): Promise<Uint8Array> => {
+    if (theirPublicKey.length !== 32) {
+      throw new Error('invalid_argument: incorrect publicKey key length for X25519')
+    }
+    const publicKey = { type: <TKeyType>'X25519', publicKeyHex: u8a.toString(theirPublicKey, 'base16') }
+    const shared = await context.agent.keyManagerSharedSecret({ secretKeyRef, publicKey })
+    return u8a.fromString(shared, 'base16')
+  }
+}
+
+export async function extractSenderEncryptionKey(
+  jwe: JWE,
+  context: IAgentContext<IResolver>,
+): Promise<Uint8Array | null> {
+  let senderKey: Uint8Array | null = null
+  const protectedHeader = decodeJoseBlob(jwe.protected)
+  if (typeof protectedHeader.skid === 'string') {
+    const senderDoc = await resolveDidOrThrow(protectedHeader.skid, context)
+    const sKey = (await context.agent.dereferenceDidUri({
```

Comment:
Why do you need to resolve the DID Doc first? Couldn't dereferenceDidUri do the job itself?

Reply:
It's because otherwise resolve() would be called too many times and currently the cache in did-resolver is not well behaved.